routes/api/contacts.js

```diff
@@ -1,80 +1,137 @@
   email: Joi.string().email(),
-  phone: Joi.string()
-}).or('name', 'email', 'phone');
+  phone: Joi.string(),
+  favorite: Joi.boolean(),
+});
 
-// Trasy
-router.get('/', (req, res) => {
-  const contacts = listContacts();
-  res.json(contacts);
+const validate = (schema, req, res, next) => {
+  const { error } = schema.validate(req.body);
+  if (error) {
+    return res.status(400).json({ message: `Validation error: ${error.details.map(detail => detail.message).join(', ')}` });
+  }
+  next();
+};
+
+const checkRequiredFields = (fields, req, res, next) => {
+  for (const field of fields) {
+    if (!req.body[field]) {
+      return res.status(400).json({ message: `Validation error: missing required field ${field}` });
+    }
+  }
+  next();
+};
+
+router.get('/', async (req, res) => {
+  try {
+    const contacts = await listContacts();
+    res.json(contacts);
+  } catch (error) {
+    res.status(500).json({ message: error.message });
+  }
 });
 
-router.get('/:id', (req, res) => {
+router.get('/:id', async (req, res) => {
   const { id } = req.params;
-  const contact = getById(id);
-
-  if (contact) {
-    res.json(contact);
-  } else {
-    res.status(404).json({ message: 'Not found' });
+  try {
+    const contact = await getById(id);
+    if (contact) {
+      res.json(contact);
+    } else {
+      res.status(404).json({ message: 'Not found' });
+    }
+  } catch (error) {
+    res.status(500).json({ message: error.message });
   }
 });
 
-router.post('/', (req, res) => {
-  const { error } = contactSchema.validate(req.body);
-
-  if (error) {
-    return res.status(400).json({ message: `Validation error: ${error.details.map(detail => detail.message).join(', ')}` });
+router.post('/', (req, res, next) => validate(baseSchema, req, res, next), (req, res, next) => checkRequiredFields(['name', 'email', 'phone'], req, res, next), async (req, res) => {
+  try {
+    const newContact = await addContact(req.body);
+    res.status(201).json(newContact);
+  } catch (error) {
+    res.status(500).json({ message: error.message });
   }
-
-  const newContact = { id: uuidv4(), ...req.body };
-  addContact(newContact);
-  res.status(201).json(newContact);
 });
 
-router.delete('/:id', (req, res) => {
+router.delete('/:id', async (req, res) => {
   const { id } = req.params;
-  const deleted = removeContact(id);
-
-  if (deleted) {
-    res.json({ message: 'Contact deleted' });
-  } else {
-    res.status(404).json({ message: 'Not found' });
+  try {
+    const success = await removeContact(id);
+    if (success) {
+      res.json({ message: 'Contact deleted' });
+    } else {
+      res.status(404).json({ message: 'Not found' });
+    }
+  } catch (error) {
+    res.status(500).json({ message: error.message });
   }
 });
 
-router.put('/:id', (req, res) => {
+router.put('/:id', (req, res, next) => validate(baseSchema, req, res, next), (req, res, next) => checkRequiredFields(['name', 'email', 'phone'], req, res, next), async (req, res) => {
```

Comment:
To tez wydaje sie skaplikowane troche. Jedno validate powinno sprawdzić wszystko :)